File: python/test/t_RandomMixture_grid1d.py
#! /usr/bin/env python

import openturns as ot

ot.TESTPREAMBLE()
ot.PlatformInfo.SetNumericalPrecision(6)
ot.ResourceMap.SetAsUnsignedInteger("RandomMixture-DefaultMaxSize", 4000000)

# Defining RandomMixture
weights = ot.Point(0)
coll = ot.DistributionCollection(0)
coll.add(ot.Gamma(0.5, 1.0))
weights.add(1.0)
coll.add(ot.Gamma(0.5, 1.0))
weights.add(1.0)
coll.add(ot.Gamma(0.5, 1.0))
weights.add(1.0)
coll.add(ot.Gamma(0.5, 1.0))
weights.add(1.0)
coll.add(ot.Gamma(1.0, 1.0))
weights.add(1.0)
distribution = ot.RandomMixture(coll, weights)
referenceDistribution = ot.Gamma(3.0, 1.0)
# Compute PDF on regular grid
N = 256
points = ot.Indices(1, N)
mean = distribution.getMean()
sigma = distribution.getStandardDeviation()
xMin = mean - 3.9 * sigma
xMax = mean + 3.9 * sigma
grid = ot.Sample()
print("distribution = ", repr(distribution))
print("range = ", distribution.getRange())
print("mean = ", distribution.getMean())
print("cov = ", distribution.getCovariance())
print("sigma = ", distribution.getStandardDeviation())
print("xMin = ", xMin)
print("xMax = ", xMax)
result, grid = distribution.computePDF(xMin, xMax, points)
print("x;PDF;reference")
for i in range(grid.getSize()):
    ref = referenceDistribution.computePDF(grid[i])
    print("%.6g;%.6g;%.6g" % (grid[i][0], result[i][0], ref))

File: lib/test/t_RandomMixture_grid1d.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class RandomMixture for grid computatioons - 1D cases
 *
 *  Copyright 2005-2019 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  PlatformInfo::SetNumericalPrecision( 6 );
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
  {
    // Defining RandomMixture
    Point weights(0);
    Collection< Distribution > coll(0);
    coll.add(Gamma(0.5, 1.0));
    weights.add(1.0);
    coll.add(Gamma(0.5, 1.0));
    weights.add(1.0);
    coll.add(Gamma(0.5, 1.0));
    weights.add(1.0);
    coll.add(Gamma(0.5, 1.0));
    weights.add(1.0);
    coll.add(Gamma(1.0, 1.0));
    weights.add(1.0);
    RandomMixture distribution(coll, weights);
    Gamma referenceDistribution(3.0, 1.0);
    // Compute PDF on regular grid
    const UnsignedInteger N = 256;
    Indices points(1, N);
    Point mean(distribution.getMean());
    Point sigma(distribution.getStandardDeviation());
    Point xMin(mean - 3.9 * sigma);
    Point xMax(mean + 3.9 * sigma);
    Sample grid;
    fullprint << "distribution = " << distribution << std::endl;
    fullprint << "distribution = " << distribution.__str__() << std::endl;
    fullprint << "range = " << distribution.getRange() << std::endl;
    fullprint << "mean = " <<  distribution.getMean() << std::endl;
    fullprint << "cov = " << distribution.getCovariance() << std::endl;
    fullprint << "sigma = " << distribution.getStandardDeviation() << std::endl;
    fullprint << "xMin = " << xMin << std::endl;
    fullprint << "xMax = " << xMax << std::endl;
    Sample result(distribution.computePDF(xMin, xMax, points, grid));
    for (UnsignedInteger i = 0; i < grid.getSize(); ++i)
      fullprint << grid[i][0] << ";" << result[i][0] << ";" << referenceDistribution.computePDF(grid[i]) << std::endl;
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}
